django/contrib/auth/hashers.py

```diff
@@ -190,8 +192,10 @@ def _load_library(self):
 
     def salt(self):
         """Generate a cryptographically secure nonce salt in ASCII."""
-        # 12 returns a 71-bit value, log_2((26+26+10)^12) =~ 71 bits
-        return get_random_string(12)
+        # Return a string of with entropy of at least self.salt_entropy
+        # A string of length N provides log_2((26+26+10)^N) bits of entropy
+        char_count = math.ceil(math.log10(math.pow(2, self.salt_entropy)) / math.log10(62))
```

Comment:
As far as I'm aware this can be simplified to:
```python
char_count = math.ceil(self.salt_entropy/math.log(62, 2))
```
also there is no need to calculate this for each `salt`, so I would move it to the cached property
```python
@cached_property
def char_count(self):
    return char_count = math.ceil(self.salt_entropy/math.log(62, 2))
```

Reply:
Is there a convention for ordering cached properties relative to other member functions? I've put this in for `BasePasswordHasher` just under the member variables.